x/disorder/writer.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2024 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package disorder
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/Jigsaw-Code/outline-sdk/x/sockopt"
+)
+
+type disorderWriter struct {
+	conn             io.Writer
+	tcpOptions       sockopt.TCPOptions
+	writesToDisorder int
+}
+
+var _ io.Writer = (*disorderWriter)(nil)
+
+// Setting number of hops to 1 will lead to data to get lost on host.
+var disorderHopN = 1
+
+func NewWriter(conn io.Writer, tcpOptions sockopt.TCPOptions, runAtPacketN int) io.Writer {
+	// TODO: Support ReadFrom.
+	return &disorderWriter{
+		conn:             conn,
+		tcpOptions:       tcpOptions,
+		writesToDisorder: runAtPacketN,
+	}
+}
+
+func (w *disorderWriter) Write(data []byte) (written int, err error) {
+	if w.writesToDisorder == 0 {
+		defaultHopLimit, err := w.tcpOptions.HopLimit()
+		if err != nil {
+			return 0, fmt.Errorf("failed to get the hop limit: %w", err)
+		}
+
+		err = w.tcpOptions.SetHopLimit(disorderHopN)
```

Comment:
The constant actually hurts readability in this case.

```suggestion
		// Setting number of hops to 1 will lead to data to get lost on host.
		err = w.tcpOptions.SetHopLimit(1)
```